src/sentry/utils/outcomes.py

```diff
@@ -122,13 +112,52 @@ def track_outcome(
         ),
     )
 
-    metrics.incr(
-        "events.outcomes",
-        skip_internal=True,
-        tags={
-            "outcome": outcome.name.lower(),
-            "reason": reason,
-            "category": category.api_name() if category is not None else "null",
-            "topic": topic_name,
-        },
+
+def track_outcome(
+    org_id: int,
+    project_id: int,
+    key_id: Optional[int],
+    outcome: Outcome,
+    reason: Optional[str] = None,
+    timestamp: Optional[datetime] = None,
+    event_id: Optional[str] = None,
+    category: Optional[DataCategory] = None,
+    quantity: Optional[int] = None,
+) -> None:
+    """
+    This is a central point to track org/project counters per incoming event.
+    NB: This should only ever be called once per incoming event, which means
+    it should only be called at the point we know the final outcome for the
+    event (invalid, rate_limited, accepted, discarded, etc.)
+
+    This sends the "outcome" message to Kafka which is used by Snuba to serve
+    data for SnubaTSDB and RedisSnubaTSDB, such as # of rate-limited/filtered
+    events.
+    """
+    track_outcome_custom_publish(
+        publish=_publish,
+        org_id=org_id,
+        project_id=project_id,
+        key_id=key_id,
+        outcome=outcome,
+        reason=reason,
+        timestamp=timestamp,
+        event_id=event_id,
+        category=category,
+        quantity=quantity,
     )
+
+
+publishers = {}
+
+
+def _publish(cluster_name: str, topic_name: str, payload: str) -> None:
+    """The publish operation used by `track_outcome`."""
+    global publishers
+
+    if cluster_name not in publishers:
+        publishers[cluster_name] = KafkaPublisher(
+            kafka_config.get_kafka_producer_cluster_options(cluster_name)
+        )
```

Comment:
Now you are using the arroyo producer sometimes and a separate non Arroyo producer which works differently if the cluster name is wrong. Also, the fact that the cluster name is sometimes randomly incorrect seems weird. I think you ensure the correct producer is passed to begin with and remove all this logic.

Reply:
Ignore this comment, I see this path doesn't apply to the billing consumer